orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/pipeline/servergroup/strategies/RollingRedBlackStrategy.groovy

```diff
@@ -86,72 +90,118 @@ class RollingRedBlackStrategy implements Strategy, ApplicationContextAware {
+      def resizeStage = newStage(
+        stage.execution,
+        resizeServerGroupStage.type,
+        "Grow to $p% Desired Size",
+        resizeContext,
+        stage,
+        SyntheticStageOwner.STAGE_AFTER
       )
+      stages << resizeStage
+
+      // an expression to grab newly deployed server group at runtime (ie. the server group being resized up)
+      def deployedServerGroupName = '${' + "#stage('${resizeStage.id}')['context']['asgName']" + '}'.toString()
+      stages.addAll(getBeforeCleanupStages(stage, stageData, source, deployedServerGroupName, p))
 
       def disableContext = baseContext + [
-          desiredPercentage           : p,
-          remainingEnabledServerGroups: 1,
-          preferLargerOverNewer       : false
+        desiredPercentage           : p,
+        remainingEnabledServerGroups: 1,
+        preferLargerOverNewer       : false
       ]
 
-      if(stageData?.delayBeforeDisableSec) {
-        def waitContext = [waitTime: stageData?.delayBeforeDisableSec]
-        stages << newStage(
-          stage.execution,
-          waitStage.type,
-          "wait",
-          waitContext,
-          stage,
-          SyntheticStageOwner.STAGE_AFTER
-        )
-      }
-
-      // TODO-AJ this has questionable behavior w/ > 1 old server group
       stages << newStage(
-          stage.execution,
-          disableClusterStage.type,
-          "Disable $p% of Traffic",
-          disableContext,
-          stage,
-          SyntheticStageOwner.STAGE_AFTER
+        stage.execution,
+        disableClusterStage.type,
+        "Disable $p% of Traffic",
+        disableContext,
+        stage,
+        SyntheticStageOwner.STAGE_AFTER
       )
-
     })
 
     return stages
   }
 
+  List<Stage> getBeforeCleanupStages(Stage parentStage,
+                                     RollingRedBlackStageData stageData,
+                                     ResizeStrategy.Source source,
+                                     String deployedServerGroupName,
+                                     int percentageComplete) {
+    def stages = []
+
+    if (stageData.getDelayBeforeCleanup()) {
+      def waitContext = [waitTime: stageData.getDelayBeforeCleanup()]
+      stages << newStage(
+        parentStage.execution,
+        waitStage.type,
+        "wait",
+        waitContext,
+        parentStage,
+        SyntheticStageOwner.STAGE_AFTER
+      )
+    }
+
+    if (stageData.pipelineBeforeCleanup) {
+      def serverGroupCoordinates = [
+        region         : source.region,
+        serverGroupName: source.serverGroupName,
+        account        : source.credentials,
+        cloudProvider  : source.cloudProvider
+      ]
+
+      def pipelineContext = [
+        pipelineApplication: stageData.pipelineBeforeCleanup.application,
+        pipelineId         : stageData.pipelineBeforeCleanup.pipelineId,
+        pipelineParameters : [
+          "deployedServerGroup": serverGroupCoordinates + [
```

Comment:
Seemed like reasonable things to pass down to the validation pipeline.

Internally we're not quite sure what these pipelines would do right now, but they're an easy extension point. Once usage picks up, common pattern might be extracted into new stages (ie. something that checks our internal metric system for deviation).